### PR TITLE
2055: TestAdapter: Test discovery fails on test projects that contain a large number of test files

### PR DESCRIPTION
**Pull request by lukedgr**
Issue #
 #2055 
##### Bug
Test discovery fails for unit test projects with 100+ test files because the command line to execute node.js test discovery from C# exceeds the windows command line length limit of 8191 characters.

##### Fix
Updated the communication between the TestFramework class in C# and find_tests.js to provide the file path for a temp file that contains the list of test files in which to discover tests.  find_tests.js then reads the contents of the provided file to get the list of test files.  This prevents exceeding the command line length maximum for projects with a large number of test files.

##### Testing
Validated that test discovery works on large projects where it failed previously

### CODE REVIEW

**Comment by armanio123** [comment]
@minestarks @amcasey This PR fixes an issue with the Test Adapter where tests discovery and execution is blocked if the tests path exceeds the command line limit.

Is this something we want to push for 15.9 ?

File: Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs [hunk 1]
```diff
@@ -35,9 +35,15 @@ public TestFramework(string vsixScriptFolder)
         {
             var testInfo = string.Empty;
             var discoverResultFile = Path.GetTempFileName();
+
+            // Create a temp file with the list of test files.
+            var testFilesString = string.Join(";", testFiles);
+            var testFilesListFile = Path.GetTempFileName();
+            File.AppendAllText(testFilesListFile, testFilesString, System.Text.Encoding.UTF8);
```
Comment (amcasey) [inline Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs]: I would probably `WriteAllText`.  I realize the file should be empty, but I think it's a clearer way to express that you intend to completely determine its contents.

File: Nodejs/Product/TestAdapter/TestFrameworks/find_tests.js [hunk 2]
```diff
@@ -6,7 +7,19 @@ try {
     process.exit(1);
 }
 try {
-    framework.find_tests(process.argv[3], process.argv[4], process.argv[5]);
+    var testFilesListInputFile = process.argv[3];
+    if (!fs.existsSync(testFilesListInputFile)) {
+        throw new Error("testFilesListInputFile '" + testFilesListInputFile + "' does not exist.");
```
Comment (amcasey) [inline Nodejs/Product/TestAdapter/TestFrameworks/find_tests.js]: It seems strange to throw within a try block.

File: Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs [hunk 3]
```diff
@@ -35,9 +35,15 @@ public TestFramework(string vsixScriptFolder)
...
+
             try
             {
-                var stdout = EvaluateJavaScript(nodeExe, string.Join(";", testFiles), discoverResultFile, logger, projectRoot);
+                var stdout = EvaluateJavaScript(nodeExe, testFilesListFile, discoverResultFile, logger, projectRoot);
```
Comment (amcasey) [inline Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs]: Is this the only call to `EvaluateJavaScript`?

Comment (armanio123) [reply Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs]: Yes it's the only call, I searched for all references in the project.

File: Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs (hunk 1 above)
Comment (armanio123) [reply Nodejs/Product/TestAdapter/TestFrameworks/TestFramework.cs]: Changed.

**Comment by armanio123** [comment]
Thanks @lukedgr for fixing the issue.

**Comment by armanio123** [comment]
Ported to 15.9. PR #2070

**Pull request merged**